### PR TITLE
keep _parent field while updating child type mapping

### LINKED ISSUE #23381: Need to respecify `_parent` when updating mapping of child type
**Elasticsearch version**: 5.2.1 docker image, 5.1.2 docker image, 5.0.1 RPM package

**Plugins installed**: []

**JVM version**: Oracle Corporation/OpenJDK 64-Bit Server VM/1.8.0_121/25.121-b13

**OS version**: Linux/4.3.0-1.el6.elrepo.x86_64/amd64

**Description of the problem including expected versus actual behavior**:
I am able to reproduce the issue presented in #15997 with Elasticsearch 5.2.1,
5.1.2 docker images, and another 5.0.1 RPM package.

Specifically:

```
PUT t 
{
  "mappings": {
    "parent": {},
    "child": {
      "_parent": {
        "type": "parent"
      }
    }
  }
}

PUT t/_mapping/child
{
  "properties": {}
}
```

results in `illegal_argument_exception`.

Again, providing `_parent` type definition in update request body worked
around the issue.

**Provide logs (if relevant)**:
```
[2017-02-27T19:51:08,706][DEBUG][o.e.a.a.i.m.p.TransportPutMappingAction] [default] failed to put mappings on indices [[[t/AK3xoZCBSCK1Wj6N81GGlA]]], type [child]
java.lang.IllegalArgumentException: The _parent field's type option can't be changed: [parent]->[null]
	at org.elasticsearch.index.mapper.ParentFieldMapper.doMerge(ParentFieldMapper.java:301) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.FieldMapper.merge(FieldMapper.java:333) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.MetadataFieldMapper.merge(MetadataFieldMapper.java:73) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.Mapping.merge(Mapping.java:96) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.index.mapper.DocumentMapper.merge(DocumentMapper.java:333) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.cluster.metadata.MetaDataMappingService$PutMappingExecutor.applyRequest(MetaDataMappingService.java:267) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.cluster.metadata.MetaDataMappingService$PutMappingExecutor.execute(MetaDataMappingService.java:230) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.cluster.service.ClusterService.executeTasks(ClusterService.java:674) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.cluster.service.ClusterService.calculateTaskOutputs(ClusterService.java:653) ~[elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.cluster.service.ClusterService.runTasks(ClusterService.java:612) [elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.cluster.service.ClusterService$UpdateTask.run(ClusterService.java:1112) [elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:527) [elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:238) [elasticsearch-5.2.1.jar:5.2.1]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:201) [elasticsearch-5.2.1.jar:5.2.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_121]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_121]
```

### PR DESCRIPTION
**Pull request by PnPie**
Be able to update child type mapping without specifying it's `_parent` field, because it should be considered as unchanged if it is not explicitly specified.
In this case, the merged mapper's `parentType` field should be `null`, so we can just merge it to keep the original one instead of throwing an exception of "trying to change it to 'null' "

Close #23381

### CODE REVIEW

**Comment by clintongormley** [comment]
@colings86 please could somebody review this

File: core/src/test/java/org/elasticsearch/index/mapper/DocumentMapperMergeTests.java [hunk 1]
```diff
@@ -235,4 +235,33 @@ public void testDoNotRepeatOriginalMapping() throws IOException {
...
+        DocumentMapper mergedMapper = initMapper.merge(updatedMapper.mapping(), false);
+
+        assertThat(mergedMapper.mappers().getMapper("_parent"), notNullValue());
+        assertThat(mergedMapper.mappers().getMapper("name"), notNullValue());
+
+        String modParentMapping = XContentFactory.jsonBuilder().startObject().startObject("child")
+            .startObject("_parent").field("type", "new_parent").endObject()
+            .endObject().endObject().string();
+        DocumentMapper modParentMapper = parser.parse("child", new CompressedXContent(modParentMapping));
+        try {
```
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/index/mapper/DocumentMapperMergeTests.java]: Replace this try-catch block with an `expectThrows(...)`  call?

```java
Exception e = expectThrows(IllegalArgumentException.class, () ->  initMapper.merge(modParentMapper.mapping(), false))
assertThat(e.getMessage(), containsString("The _parent field's type option can't be changed:"));
```

File: core/src/test/java/org/elasticsearch/index/mapper/DocumentMapperMergeTests.java [hunk 2]
```diff
@@ -235,4 +235,33 @@ public void testDoNotRepeatOriginalMapping() throws IOException {
...
+            .endObject().endObject().string();
+        DocumentMapper initMapper = parser.parse("child", new CompressedXContent(initMapping));
+
+        String updatedMapping = XContentFactory.jsonBuilder().startObject().startObject("child")
+            .startObject("properties")
+            .startObject("name").field("type", "text").endObject()
+            .endObject().endObject().endObject().string();
+        DocumentMapper updatedMapper = parser.parse("child", new CompressedXContent(updatedMapping));
+        DocumentMapper mergedMapper = initMapper.merge(updatedMapper.mapping(), false);
+
```
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/index/mapper/DocumentMapperMergeTests.java]: Maybe also add a merge  with current parent field and a new field?

```java
String updatedMapping = XContentFactory.jsonBuilder().startObject().startObject("child")
   .startObject("_parent").field("type", "parent").endObject()
   .startObject("properties")
   .startObject("field2").field("type", "text").endObject()
   .endObject().endObject().endObject().string();
 DocumentMapper updatedMapper = parser.parse("child", new CompressedXContent(updatedMapping));
 DocumentMapper mergedMapper = initMapper.merge(updatedMapper.mapping(), false);
```

**Review by martijnvg — COMMENTED** [review]
Thanks @PnPie! This looks good. I left two comments regarding the test.

**Comment by PnPie** [comment]
Thanks for your comments @martijnvg, I added a merge case with current parent field while I found a new problem: I cannot find the `_parent` field with this assertion didn't pass
`assertThat(mergedMapper.mappers().getMapper("_parent"), notNullValue());`

And then I found that the parent field name of a document is in format: `_parent#[parent type]`, so in the test it should be `_parent#parent`, if a new mapper without parent field, it's name is just `_parent`, and when we do the merge we firstly call the `super.doMerge` method of `ParentFieldMapper` and it always replaces the `fieldType` with the new one's, so in consequence the parent field name is changed to the new one's.
https://github.com/PnPie/elasticsearch/blob/346889a93bb18cda93736a8076aef198b8f86fe1/core/src/main/java/org/elasticsearch/index/mapper/ParentFieldMapper.java#L298
So previously after this we can find the `_parent` field with
`assertThat(mergedMapper.mappers().getMapper("_parent"), notNullValue());`
but in fact the parent field shouldn't be changed and it should be `_parent#parent`, so I modified in `ParentFieldMapper`'s `doMerge` method and make it keep the original parent field name.

And I also add a test for trying to add parent field to an exiting mapper without parent.

**Comment by martijnvg** [comment]
Cool, looks good @PnPie. I'll merge it in soon and backport to 5.x branch.

**Pull request merged**